### ISSUE TITLE
Building 1.5.3-1.5.5 fails on EnterpriseDB 11.6.13

### ISSUE THREAD

**Issue by berendiwema** [tada/pljava#260]
Building any 1.5.3, 1.5.4 and 1.5.5 fails on a Postgres 11 based environment on CentOS 7. 
1.5.2 builds are OK.

Building with apache-maven-3.6.3, command mvn -Pwnosign clean install
```
Starting Core 0 with 53 source files...
[ERROR] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/PgSavepoint.c: In function 'Java_org_postgresql_pljava_internal_PgSavepoint__1release':
[ERROR] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/PgSavepoint.c:143:10: error: 'ReleaseCurrentSubTransaction' undeclared (first use in this function)
[ERROR]    unwind(ReleaseCurrentSubTransaction, xid, nestLevel);
[ERROR]           ^
[ERROR] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/PgSavepoint.c:143:10: note: each undeclared identifier is reported only once for each function it appears in
[ERROR] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/SubXactListener.c: In function 'subXactCB':
[WARNING] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/SubXactListener.c:45:2: warning: enumeration value 'SUBXACT_EVENT_PRE_COMMIT_SUB' not handled in switch [-Wswitch]
[ERROR]   switch(event)
[ERROR]   ^
[ERROR] In file included from /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/type/UDT.c:19:0:
[ERROR] /usr/edb/as11/include/server/libpq/pqformat.h: In function 'pq_writeint8':
[WARNING] /usr/edb/as11/include/server/libpq/pqformat.h:53:11: warning: conversion to 'int' from 'long unsigned int' may alter its value [-Wconversion]
[ERROR]   buf->len += sizeof(uint8);
[ERROR]            ^
[ERROR] /usr/edb/as11/include/server/libpq/pqformat.h: In function 'pq_writeint16':
[WARNING] /usr/edb/as11/include/server/libpq/pqformat.h:67:11: warning: conversion to 'int' from 'long unsigned int' may alter its value [-Wconversion]
[ERROR]   buf->len += sizeof(uint16);
[ERROR]            ^
[ERROR] /usr/edb/as11/include/server/libpq/pqformat.h: In function 'pq_writeint32':
[WARNING] /usr/edb/as11/include/server/libpq/pqformat.h:81:11: warning: conversion to 'int' from 'long unsigned int' may alter its value [-Wconversion]
[ERROR]   buf->len += sizeof(uint32);
[ERROR]            ^
[ERROR] /usr/edb/as11/include/server/libpq/pqformat.h: In function 'pq_writeint64':
[WARNING] /usr/edb/as11/include/server/libpq/pqformat.h:95:11: warning: conversion to 'int' from 'long unsigned int' may alter its value [-Wconversion]
[ERROR]   buf->len += sizeof(uint64);
[ERROR]            ^
[ERROR] /usr/edb/as11/include/server/libpq/pqformat.h: In function 'pq_writestring':
[WARNING] /usr/edb/as11/include/server/libpq/pqformat.h:111:21: warning: conversion to 'int' from 'size_t' may alter its value [-Wconversion]
[ERROR]   int   slen = strlen(str);
[ERROR]                      ^
[WARNING] /usr/edb/as11/include/server/libpq/pqformat.h:116:16: warning: conversion to 'int' from 'size_t' may alter its value [-Wconversion]
[ERROR]    slen = strlen(p);
[ERROR]                 ^
[ERROR] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/XactListener.c: In function 'xactCB':
[WARNING] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/XactListener.c:26:2: warning: enumeration value 'XACT_EVENT_PARALLEL_COMMIT' not handled in switch [-Wswitch]
[WARNING] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/XactListener.c:26:2: warning: enumeration value 'XACT_EVENT_PARALLEL_COMMIT' not handled in switch [-Wswitch]
[ERROR]   switch(event)
[ERROR]   ^
[WARNING] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/XactListener.c:26:2: warning: enumeration value 'XACT_EVENT_PARALLEL_ABORT' not handled in switch [-Wswitch]
[WARNING] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/XactListener.c:26:2: warning: enumeration value 'XACT_EVENT_PRE_COMMIT' not handled in switch [-Wswitch]
[WARNING] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/XactListener.c:26:2: warning: enumeration value 'XACT_EVENT_PARALLEL_PRE_COMMIT' not handled in switch [-Wswitch]
[WARNING] /usr/local/src/pljava-1_5_3/pljava-so/src/main/c/XactListener.c:26:2: warning: enumeration value 'XACT_EVENT_PRE_PREPARE' not handled in switch [-Wswitch]
[INFO] 53 files were compiled.                                      
[INFO] 53 files were compiled.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for PostgreSQL PL/Java 1.5.3:
[INFO] 
[INFO] PostgreSQL PL/Java ................................. SUCCESS [ 19.003 s]
[INFO] PL/Java API ........................................ SUCCESS [ 10.252 s]
[INFO] PL/Java backend Java code .......................... SUCCESS [  3.925 s]
[INFO] PL/Java backend native code ........................ FAILURE [ 17.900 s]
[INFO] PL/Java Deploy ..................................... SKIPPED
[INFO] PL/Java Ant tasks .................................. SKIPPED
[INFO] PL/Java examples ................................... SKIPPED
[INFO] PL/Java packaging .................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:01 min
[INFO] Finished at: 2020-04-14T09:14:04Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.maven-nar:nar-maven-plugin:3.2.3:nar-compile (default-nar-compile) on project pljava-so: NAR: Compile failed: gcc failed with return code 1 -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :pljava-so
```

**Comment by jcflack** [comment]
Hello,

I am unable to reproduce this error, building 1.5.5 with PostgreSQL 11.7.

The actual error in what you have posted is the first one shown, that `ReleaseCurrentSubTransaction` is undeclared. However, that is declared in the `access/xact.h` file included in PostgreSQL's development headers, and the PL/Java file `PgSavepoint.c` does contain an `#include` for that file.

So it is difficult for me to picture what is happening in your build. Could you please report the full version of PostgreSQL being used, and which C compiler?

Also, could you include the output of `pg_config --includedir-server` and

    fgrep ReleaseCurrentSubTransaction $(pg_config --includedir-server)/access/xact.h

Thank you!

**Comment by berendiwema** [comment]
I am building against  EnterpriseDB 11.7 (might be the culprit... not sure)

g++ version:
```
  Installing : cpp-4.8.5-39.el7.x86_64                                                                                                                                      
  Installing : libgomp-4.8.5-39.el7.x86_64                                                                                                                                  
  Installing : kernel-headers-3.10.0-1127.el7.x86_64                                                                                                                        
  Installing : glibc-headers-2.17-307.el7.1.x86_64                                                                                                                          
  Installing : glibc-devel-2.17-307.el7.1.x86_64                                                                                                                             
  Installing : gcc-4.8.5-39.el7.x86_64                                                                                                                                     
  Installing : libstdc++-devel-4.8.5-39.el7.x86_64                                                                                                                         
  Installing : gcc-c++-4.8.5-39.el7.x86_64               
```

Postgres version info:
```
sh-4.2$ psql --version
psql (EnterpriseDB) 11.6.13
enterprisedb@acs-db-master-0 ~]$ pg_config --includedir-server
/usr/edb/as11/include/server
```

fgrep output:
```
[enterprisedb@acs-db-master-0 ~]$ fgrep ReleaseCurrentSubTransaction $(pg_config --includedir-server)/access/xact.h
extern void ReleaseCurrentSubTransactionEx(bool edb_allow_parallel);
#define ReleaseCurrentSubTransaction() \
	ReleaseCurrentSubTransactionEx(false)
extern void RollbackAndReleaseCurrentSubTransaction(void);

```

I haven't been able to go in depth in the code and examine the difference between 1.5.2 and 1.5.3 as of yet, however that would be my next step if you don't recognize the issue either. 
Thanks a bunch for your help thusfar!

**Comment by jcflack** [comment]
I think I see it. The corresponding [line in upstream PostgreSQL][l389] is just:

```
extern void ReleaseCurrentSubTransaction(void);
```

EnterpriseDB made a new version  of that function, `ReleaseCurrentSubTransactionEx`, that takes an `edb_allow_parallel` argument. In place of the original function, they just left behind a compiler macro that turns `ReleaseCurrentSubTransaction()` into `ReleaseCurrentSubTransactionEx(false)`.

But PL/Java is doing [this][l143]:

```
		unwind(ReleaseCurrentSubTransaction, xid, nestLevel);
```

that is, not calling the function, but passing the function's address to `unwind()`. That's because `unwind` is the exact same process [21 lines later][l164], except for the function to use:

```
		unwind(RollbackAndReleaseCurrentSubTransaction, xid, nestLevel);
```

After EnterpriseDB made the function go away and left only the macro, there is nothing there to take the address of. In general, making an API change like that, they should leave behind an actual function also, which would just call the new one, passing `false`.

Could you check a couple of things?

* Did they also do the same thing with `RollbackAndReleaseCurrentSubTransaction`?
* Is there some EnterpriseDB-specific macro that could be used in an `#if` to make a conditional workaround?
* Can you easily determine which versions of EnterpriseDB contain this change?

Thanks!

[l389]: https://github.com/postgres/postgres/blob/34f805c/xact.h#L389
[l143]: https://github.com/tada/pljava/blob/213f897/pljava-so/src/main/c/PgSavepoint.c#L143
[l164]: https://github.com/tada/pljava/blob/213f897/pljava-so/src/main/c/PgSavepoint.c#L164

**Comment by berendiwema** [comment]
That was quick :)

1) Did they also do the same thing with RollbackAndReleaseCurrentSubTransaction?
fgrep tells me this:
```
[enterprisedb@acs-db-master-0 ~]$ fgrep RollbackAndReleaseCurrentSubTransaction $(pg_config --includedir-server)/access/xact.h
extern void RollbackAndReleaseCurrentSubTransaction(void);
```
Which looks the same as https://github.com/postgres/postgres/blob/master/src/include/access/xact.h#L412

2) Is there some EnterpriseDB-specific macro that could be used in an #if to make a conditional workaround?

In xact.h, there is an enterprisedb-specific enum, is that something which could help?
```
typedef enum
{
[....]
} EDBXactEvent;
```

3) On EnterpriseDB 10, the customizations look the same:
```
[enterprisedb@edb10 ~]$ fgrep ReleaseCurrentSubTransaction $(pg_config --includedir-server)/access/xact.h
extern void ReleaseCurrentSubTransactionEx(bool edb_allow_parallel);
#define ReleaseCurrentSubTransaction() \
	ReleaseCurrentSubTransactionEx(false)
extern void RollbackAndReleaseCurrentSubTransaction(void);
```

If you like, I can check EDB 9 and 12 as well this week.

**Comment by jcflack** [comment]
A preprocessor conditional can only depend on stuff known to the preprocessor, like macros.

It's possible that if you open an issue with EDB, they'll be able to tell you exactly which releases contain that code, and a good way to recognize them at compile time. (I could always go by the existence of `ReleaseCurrentSubTransaction` as a macro, but I would prefer something more explicit.)

**Comment by berendiwema** [comment]
Alright, I will open an issue with EDB to check since when they added that code, and ask for a preprocessor condition. I will get back to you as soon as I hear from them.

**Comment by jcflack** [comment]
I have just pushed a branch `workaround/REL1_5_STABLE/issue260`, in case you would like to check whether it builds successfully with EDB. I think it will resolve the issue reported here, and I mean to include it in the upcoming release.

**Comment by berendiwema** [comment]
Thanks jcflack, I  will check it out right away.

I haven't received anything concrete from EDB as of yet, except that they are escalating the issue. I think I took the wrong path going through support instead of contacts.

**Comment by berendiwema** [comment]
So I've just tried another build, which was not successful unfortunately.
Looks like it's exiting on compiler warnings. See: https://gist.github.com/berendiwema/5dd993dba78f6e6cb3a31a0a0dbf0674

I assume these warnings are there for a reason, is there anything I can try to suppress them?

**Comment by jcflack** [comment]
You're right, the 1.5.x builds produce a great many warnings, which makes it hard to spot where the error is, if there is one. In this case, it's [here][L1509].

That's something people commonly encounter, and simply means the OpenSSL development package needs to be installed. (And it gets frequently reported as a build problem here, because so many warnings make it hard to see the error.)

There are some [tips on the wiki][tips] for build troubleshooting, including command options that can reduce (but not completely eliminate) the warnings, so problems are easier to spot.

1.6 will have a much improved build process (developed this summer) without the excess warnings, but it isn't released yet. (And its branch won't build for you yet, because I haven't copied this EDB workaround there yet.)

[L1509]: https://gist.github.com/berendiwema/5dd993dba78f6e6cb3a31a0a0dbf0674#file-gistfile1-txt-L1509
[tips]: https://github.com/tada/pljava/wiki/Build-tips

**Comment by berendiwema** [comment]
Ah man, I did encounter that issue before and solved it as well... been a while. Thanks for checking. I'll give it another go right away.

**Comment by berendiwema** [comment]
@jcflack: Success!
`[INFO] Executed tasks
[INFO] 
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ pljava-packaging ---
[INFO] Installing /var/lib/edb/pljava/pljava-packaging/pom.xml to /var/lib/edb/.m2/repository/org/postgresql/pljava-packaging/1.5.6-SNAPSHOT/pljava-packaging-1.5.6-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for PostgreSQL PL/Java 1.5.6-SNAPSHOT:
[INFO] 
[INFO] PostgreSQL PL/Java ................................. SUCCESS [  0.921 s]
[INFO] PL/Java API ........................................ SUCCESS [  1.662 s]
[INFO] PL/Java backend Java code .......................... SUCCESS [  2.169 s]
[INFO] PL/Java backend native code ........................ SUCCESS [ 21.445 s]
[INFO] PL/Java Deploy ..................................... SUCCESS [  1.336 s]
[INFO] PL/Java Ant tasks .................................. SUCCESS [  2.056 s]
[INFO] PL/Java examples ................................... SUCCESS [  0.523 s]
[INFO] PL/Java packaging .................................. SUCCESS [  1.153 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  31.892 s
[INFO] Finished at: 2020-09-30T14:53:42Z
[INFO] ------------------------------------------------------------------------`

**Comment by jcflack** [comment]
Great! I'll make a pull request out of that workaround.

Don't feel bad about missing that error. The torrent of warnings has been a headache for everybody for years. It just took a Google Summer of Code sponsorship this year to produce enough round tuits to do something about it.

**Comment by berendiwema** [comment]
Awesome, thank you for helping out. I'll be including the next release of pl/java in our platform 👍

**Comment by jcflack** [comment]
Workaround included in 1.5.6, which is now released (and will also be included in the upcoming 1.6.0).

Are you able to give any description of your platform? I'm just curious; it's nice to have a sense of what the stuff is being used for. Sometimes it also helps judge what development directions would be useful.

**Comment by berendiwema** [comment]
Awesome, thanks @jcflack !

Sure; we run an OpenShift platform with over 100 different projects on it, all with the same components. A major application we deploy is Pega, which makes use of the PL/Java engine to speed up some processes!

**Comment by jcflack** [comment]
Ah, ok.  I well remember getting a bunch of help from Pega's Ken Olson when getting the first 1.5 releases into shape.

Anyway, this workaround is in both recently-released PL/Java versions, 1.5.6 two weeks ago, and 1.6.0 today for applications with PostgreSQL >= 9.5 and Java >=9. I'll mark this resolved.
